server/session.go

```diff
@@ -0,0 +1,194 @@
+		tests:            []testEntry{},
+	}
+	return session
+}
+
+type testEntry struct {
+	test    Test
+	deleted bool
+}
+
+type sessionImpl struct {
+	name             string
+	version          pb.Session_Version
+	observerRegistry GrpcObserverRegistry
+	token            TokenGenerator
+
+	mu    sync.Mutex
+	keys  map[string]int
+	tests []testEntry
+}
+
+func (s *sessionImpl) GetName() string {
+	return s.name
+}
+
+func (s *sessionImpl) GetVersion() pb.Session_Version {
+	return s.version
+}
+
+func (s *sessionImpl) GetReport() *pb.ReportSessionResponse {
+	result := pb.ReportSessionResponse_PASSED
+	for _, entry := range s.tests {
+		if entry.deleted {
+			continue
+		}
+		test := entry.test
+		issue := test.GetIssue()
+		if issue == nil {
+			continue
+		}
+		// TODO: Add severity handling
+		if issue.GetType() == pb.Issue_INCORRECT_CONFIRMATION {
+			result = pb.ReportSessionResponse_FAILED
+			continue
+		}
+		if result != pb.ReportSessionResponse_FAILED {
+			result = pb.ReportSessionResponse_INCOMPLETE
+		}
+	}
+
+	testRuns := []*pb.TestRun{}
+	for _, entry := range s.tests {
+		if entry.deleted {
+			continue
+		}
+		testRuns = append(testRuns, TestRunProto(entry.test))
+	}
+	return &pb.ReportSessionResponse{
+		Result:   result,
+		TestRuns: testRuns,
+	}
+}
+
+func (s *sessionImpl) ListTests(in *pb.ListTestsRequest) (*pb.ListTestsResponse, error) {
+	start, err := s.token.GetIndex(in.GetPageToken())
+	pageSize := in.GetPageSize()
+	if pageSize == 0 {
+		pageSize = 10
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	offset := 0
+	tests := []*pb.Test{}
+	for _, entry := range s.tests[start:] {
+		offset++
+		if entry.deleted {
+			continue
+		}
+		tests = append(tests, TestProto(entry.test))
+		if len(tests) >= int(pageSize) {
+			break
+		}
+	}
+
+	nextToken := ""
+	if start+offset < len(s.tests) {
+		nextToken = s.token.ForIndex(start + offset)
+	}
+
+	return &pb.ListTestsResponse{Tests: tests, NextPageToken: nextToken}, nil
+}
+
+func (s *sessionImpl) DeleteTest(name string) (*empty.Empty, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for n, i := range s.keys {
+		if strings.HasSuffix(name, n) && !s.tests[i].deleted {
+			s.tests[i] = testEntry{test: s.tests[i].test, deleted: true}
+
+			test := s.tests[i].test
+			if _, ok := test.(UnaryObserver); ok {
+				s.observerRegistry.DeleteUnaryObserver(test.GetName())
+			}
+			if _, ok := test.(StreamRequestObserver); ok {
+				s.observerRegistry.DeleteStreamRequestObserver(test.GetName())
+			}
+			if _, ok := test.(StreamResponseObserver); ok {
+				s.observerRegistry.DeleteStreamResponseObserver(test.GetName())
+			}
+
+			return &empty.Empty{}, nil
+		}
+	}
+
+	return nil, status.Errorf(
+		codes.NotFound,
+		"A test with name %s not found.", name)
+}
+
+func (s *sessionImpl) RegisterTests(tests []Test) {
+	for _, test := range tests {
+		i := len(s.tests)
+		s.tests = append(s.tests, testEntry{test: test})
+		s.keys[test.GetName()] = i
+		if obs, ok := test.(UnaryObserver); ok {
```

Comment:
Same here. A type switch might be be better

Reply:
I don't think a type switch will be applicable here since the type switch will only run the first type that it matches and fallthroughs in type switches dont work.

https://play.golang.org/p/gSlDbQGzVqX

Reply:
Ah I missed the fact that `A Test will also implement at least one of` the observers. Cool